### PR TITLE
Remove globals, add context propagation to CLI

### PR DESCRIPTION
**Pull request by stefanmcshane**
## Description

Goal of this was to remove globals and support context propagation in order to bring in better telemetry for CLI actions, as well as support multiple profiles/projects/clusters without continuously logging out and in.

## Summary of work
- remove globals
- add context propagation
- update to address linter without changing logic where possible, incase existing logic relies on not checking errors
- increase cyclo complexity because a lot of this code needs refactored to be easier to understand

### CODE REVIEW

File: cli/cmd/errors.go [hunk 1]
```diff
@@ -18,10 +19,19 @@ var (
...
+func checkLoginAndRun(ctx context.Context, args []string, runner func(ctx context.Context, user *types.GetAuthenticatedUserResponse, client api.Client, cliConfig config.CLIConfig, args []string) error) error {
+	cliConf, err := config.InitAndLoadConfig()
+	if err != nil {
+		return fmt.Errorf("error loading porter config: %w", err)
+	}
+
+	client := api.NewClientWithConfig(ctx, api.NewClientInput{
+		BaseURL:        fmt.Sprintf("%s/api", cliConf.Host),
+		BearerToken:    cliConf.Token,
+		CookieFileName: "cookie.json",
```
Comment (d-g-town) [inline cli/cmd/errors.go:31]: `NewClientWithConfig` doesn't do anything with `cookie.json`

File: cli/cmd/open.go [hunk 2]
```diff
@@ -14,15 +16,26 @@ var openCmd = &cobra.Command{
...
+		if err != nil {
+			color.New(color.FgRed).Fprintf(os.Stderr, "error loading porter config: %s\n", err) //nolint:errcheck,gosec // do not want to change logic of CLI. New linter error
+			os.Exit(1)
+		}
 
-		user, err := client.AuthCheck(context.Background())
+		client := api.NewClientWithConfig(ctx, api.NewClientInput{
+			BaseURL:        fmt.Sprintf("%s/api", cliConf.Host),
+			BearerToken:    cliConf.Token,
+			CookieFileName: "cookie.json",
```
Comment (d-g-town) [inline cli/cmd/open.go:29]: same thing about `NewClientWithConfig` and `cookie.json`

File: api/client/api.go [hunk 3]
```diff
@@ -25,7 +26,52 @@ type Client struct {
...
+
+	// BearerToken uses a JWT to authenticate with the Porter API. If both BearerToken and CookieFileName are specified, BearerToken will be used
+	BearerToken string
+
+	// CloudflareToken allows for authenticating with a Porter API behind Cloudflare Zero Trust. If not specified, we will check PORTER_CF_ACCESS_TOKEN for a token.
+	// If one is found, it will be added to all API calls.
+	CloudflareToken string
+
+	// // CLIConfig contains all config read from flags, environment variables, or porter.yaml config. This is used to automatically pull hosts, projectIDs, clusterIDs etc. in API calls
+	// CLIConfig config.CLIConfig
```
Comment (d-g-town) [inline api/client/api.go:53]: is the plan to uncomment this before this merges?

File: cli/cmd/porter_app/apply.go [hunk 4]
```diff
@@ -18,16 +18,23 @@ import (
...
 }
 
-func CreateApplicationDeploy(client *api.Client, worker *switchboardWorker.Worker, app *Application, applicationName string, cliConf *config.CLIConfig) ([]*switchboardTypes.Resource, error) {
+// CreateApplicationDeploy creates everything needed to deploy a porter app
+func CreateApplicationDeploy(ctx context.Context, client api.Client, worker *switchboardWorker.Worker, app *Application, applicationName string, cliConf config.CLIConfig) ([]*switchboardTypes.Resource, error) {
+	err := cliConf.ValidateCLIEnvironment()
+	if err != nil {
+		errMsg := composePreviewMessage("porter CLI is not configured correctly", Error)
+		return nil, fmt.Errorf("%s: %w", errMsg, err)
+	}
```
Comment (d-g-town) [inline cli/cmd/porter_app/apply.go:33]: why add this here?

**Review by d-g-town — APPROVED** [review]

File: cli/cmd/porter_app/apply.go (hunk 4 above)
Comment (stefanmcshane) [reply cli/cmd/porter_app/apply.go:33]: This check was being performed a few levels deeper using the global cliConf. This just moves the evaluation of environment variables closer to the initial call

**Pull request merged**